grpc-server/tinkerbell.go

```diff
@@ -2,47 +2,219 @@ package grpcserver
 
-	exec "github.com/tinkerbell/tink/executor"
+	"github.com/tinkerbell/tink/db"
 	pb "github.com/tinkerbell/tink/protos/workflow"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
+var workflowData = make(map[string]int)
+
 // GetWorkflowContexts implements tinkerbell.GetWorkflowContexts
 func (s *server) GetWorkflowContexts(req *pb.WorkflowContextRequest, stream pb.WorkflowSvc_GetWorkflowContextsServer) error {
-	return exec.GetWorkflowContexts(req, stream, s.db)
+	wfs, err := getWorkflowsForWorker(s.db, req.WorkerId)
+	if err != nil {
+		return err
+	}
+	for _, wf := range wfs {
+		wfContext, err := s.db.GetWorkflowContexts(context.Background(), wf)
+		if err != nil {
+			return status.Errorf(codes.Aborted, "invalid workflow %s found for worker %s", wf, req.WorkerId)
+		}
+		if isApplicableToSend(context.Background(), wfContext, req.WorkerId, s.db) {
+			if err := stream.Send(wfContext); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
 }
 
 // GetWorkflowContextList implements tinkerbell.GetWorkflowContextList
 func (s *server) GetWorkflowContextList(context context.Context, req *pb.WorkflowContextRequest) (*pb.WorkflowContextList, error) {
-	return exec.GetWorkflowContextList(context, req, s.db)
+	wfs, err := getWorkflowsForWorker(s.db, req.WorkerId)
+	if err != nil {
+		return nil, err
+	}
+	wfContexts := []*pb.WorkflowContext{}
+	for _, wf := range wfs {
+		wfContext, err := s.db.GetWorkflowContexts(context, wf)
+		if err != nil {
+			return nil, status.Errorf(codes.Aborted, "Invalid workflow %s found for worker %s", wf, req.WorkerId)
+		}
+		wfContexts = append(wfContexts, wfContext)
+	}
+	return &pb.WorkflowContextList{
+		WorkflowContexts: wfContexts,
+	}, nil
 }
 
 // GetWorkflowActions implements tinkerbell.GetWorkflowActions
 func (s *server) GetWorkflowActions(context context.Context, req *pb.WorkflowActionsRequest) (*pb.WorkflowActionList, error) {
-	return exec.GetWorkflowActions(context, req, s.db)
+	wfID := req.GetWorkflowId()
+	if len(wfID) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "workflow_id is invalid")
+	}
+	return getWorkflowActions(context, s.db, wfID)
 }
 
 // ReportActionStatus implements tinkerbell.ReportActionStatus
 func (s *server) ReportActionStatus(context context.Context, req *pb.WorkflowActionStatus) (*pb.Empty, error) {
-	return exec.ReportActionStatus(context, req, s.db)
+	wfID := req.GetWorkflowId()
+	if len(wfID) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "workflow_id is invalid")
+	}
+	if len(req.GetTaskName()) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "task_name is invalid")
+	}
+	if len(req.GetActionName()) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "action_name is invalid")
+	}
+	fmt.Printf("Received action status: %s\n", req)
+	wfContext, err := s.db.GetWorkflowContexts(context, wfID)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "workflow context not found for workflow %s", wfID)
```

Comment:
this error statement seems to be valid with my above comment on same point. May be we can refine the statement and it here as well.